go/test/endtoend/vtgate/buffer/buffer_test.go

```diff
@@ -0,0 +1,453 @@
+		if err != nil {
+			c.errors++
+			println(err.Error())
+		}
+		c.rpcs++
+		// If notifications are requested, check if we already executed the
+		// required number of successful RPCs.
+		// Use >= instead of == because we can miss the exact point due to
+		// slow thread scheduling.
+		c.notifyLock.Lock()
+		if c.notifyAfterNSuccessfulRpcs != 0 && c.rpcs >= (c.notifyAfterNSuccessfulRpcs+c.rpcsSoFar) {
+			c.waitForNotification <- true
+			c.notifyAfterNSuccessfulRpcs = 0
+		}
+		c.notifyLock.Unlock()
+		time.Sleep(10 * time.Millisecond)
+	}
+	wg.Done()
+}
+
+func (c *threadParams) setNotifyAfterNSuccessfulRpcs(n int) {
+	println("before main thread lock")
+	c.notifyLock.Lock()
+	println("after main thread lock")
+	c.notifyAfterNSuccessfulRpcs = n
+	c.rpcsSoFar = c.rpcs
+	c.notifyLock.Unlock()
+}
+
+func (c *threadParams) stop() {
+	c.quit = true
+}
+
+func readExecute(c *threadParams, conn *mysql.Conn) error {
+	_, err := conn.ExecuteFetch(fmt.Sprintf("SELECT * FROM buffer WHERE id = %d", criticalReadRowID), 1000, true)
+	return err
+}
+
+func updateExecute(c *threadParams, conn *mysql.Conn) error {
+	attempts := c.i
+	c.i++
+	conn.ExecuteFetch("begin", 1000, true)
+	// Do not use a bind variable for "msg" to make sure that the value shows
+	// up in the logs.
+	_, err := conn.ExecuteFetch(fmt.Sprintf("UPDATE buffer SET msg='update %d' WHERE id = %d", attempts, updateRowID), 1000, true)
+	// Sleep between [0, 1] seconds to prolong the time the transaction is in
+	// flight. This is more realistic because applications are going to keep
+	// their transactions open for longer as well.
+	time.Sleep(time.Duration(rand.Int31n(1000)) * time.Millisecond)
+	if err == nil {
+		fmt.Printf("update %d affected and %d notification value", attempts, c.notifyAfterNSuccessfulRpcs)
+		_, err = conn.ExecuteFetch("commit", 1000, true)
+		if err != nil {
+			_, errRollback := conn.ExecuteFetch("rollback", 1000, true)
+			if errRollback != nil {
+				fmt.Print("Error in rollback", errRollback.Error())
+			}
+			c.commitErrors++
+			if c.commitErrors > 1 {
+				return err
+			}
+			fmt.Printf("UPDATE %d failed during COMMIT. This is okay once because we do not support buffering it. err: %s", attempts, err.Error())
+		}
+	}
+	if err != nil {
+		_, errRollback := conn.ExecuteFetch("rollback", 1000, true)
+		if errRollback != nil {
+			fmt.Print("Error in rollback", errRollback.Error())
+		}
+		// if !commitStarted {
+		// 	fmt.Printf("UPDATE %d failed before COMMIT. This should not happen.Re-raising exception.", attempts)
+		// 	err = errors.New("UPDATE failed before COMMIT")
+		// 	return err
+		// }
+		c.commitErrors++
+		if c.commitErrors > 1 {
+			return err
+		}
+		fmt.Printf("UPDATE %d failed during COMMIT. This is okay once because we do not support buffering it. err: %s", attempts, err.Error())
+	}
+	return nil
+}
+
+func (c *threadParams) getCommitErrors() int {
+	return c.commitErrors
+}
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+	m.Run()
+}
```

Comment:
I think we don't need `TestMain` in this file.

Reply:
done